File: BE/product-service/src/main/java/com/ecommerce/productservice/entity/Product.java
package com.ecommerce.productservice.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @Column(name = "name", nullable = false, length = 500)
    private String name;

    @Column(name = "slug", nullable = false, unique = true, length = 500)
    private String slug;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "brand")
    private String brand;

    @Column(name = "base_price", nullable = false, precision = 19, scale = 4)
    @Builder.Default
    private BigDecimal basePrice = BigDecimal.ZERO;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 50)
    @Builder.Default
    private ProductStatus status = ProductStatus.ACTIVE;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "attributes", columnDefinition = "jsonb")
    private Map<String, Object> attributes;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "images", columnDefinition = "jsonb")
    @Builder.Default
    private List<String> images = new ArrayList<>();

    @Column(name = "es_synced", nullable = false)
    @Builder.Default
    private boolean esSynced = false;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Column(name = "avg_rating", nullable = false, precision = 3, scale = 2)
    @Builder.Default
    private BigDecimal avgRating = BigDecimal.ZERO;

    @Column(name = "review_count", nullable = false)
    @Builder.Default
    private Integer reviewCount = 0;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    private List<Sku> skus = new ArrayList<>();
}

File: FE/src/types/product.ts
export interface Category {
  id: string
  parentId?: string
  name: string
  slug: string
  description?: string
  imageUrl?: string
  children?: Category[]
}

export interface Sku {
  id: string
  skuCode: string
  variantName?: string
  attributes: Record<string, string>
  price: number
  costPrice?: number
  weightGrams?: number
  active: boolean
}

export interface SkuRequest {
  skuCode: string
  variantName?: string
  attributes?: Record<string, string>
  price: number
  costPrice?: number
  weightGrams?: number
  active: boolean
}

export interface Product {
  id: string
  categoryId?: string
  categoryName?: string
  name: string
  slug: string
  description?: string
  brand?: string
  basePrice: number
  status: 'ACTIVE' | 'INACTIVE' | 'OUT_OF_STOCK' | 'DELETED'
  attributes: Record<string, unknown>
  images: string[]
  skus: Sku[]
  createdAt: string
  avgRating?: number
  reviewCount?: number
}

export interface ProductSearchParams {
  q?: string
  category?: string
  minPrice?: number
  maxPrice?: number
  brand?: string
  page?: number
  size?: number
  sort?: string
}
